test/tls12groupselection_test.c

```diff
@@ -0,0 +1,201 @@
+        }
+        TEST_true_or_end(SSL_CTX_set_min_proto_version(client_ctx, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_CTX_set_min_proto_version(server_ctx, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_CTX_set_max_proto_version(client_ctx, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_CTX_set_max_proto_version(server_ctx, TLS1_2_VERSION));
+        if (current_test_vector->preference == SERVER_PREFERENCE)
+            SSL_CTX_set_options(server_ctx, SSL_OP_CIPHER_SERVER_PREFERENCE);
+    }
+    /* Creation of the SSL objects */
+    if (!TEST_true(create_ssl_objects(server_ctx, client_ctx,
+                                      &serverssl, &clientssl,
+                                      NULL, NULL)))
+        goto end;
+
+    /* Customization of the SSL objects */
+    if (ssl_or_ctx == WORK_ON_SSL_OBJECT) {
+        if (current_test_vector->client_groups != NULL)
+            TEST_true_or_end(SSL_set1_groups_list(clientssl, current_test_vector->client_groups));
+        if (current_test_vector->server_groups != NULL)
+            TEST_true_or_end(SSL_set1_groups_list(serverssl, current_test_vector->server_groups));
+        TEST_true_or_end(SSL_set_min_proto_version(clientssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_min_proto_version(serverssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_max_proto_version(clientssl, TLS1_2_VERSION));
+        TEST_true_or_end(SSL_set_max_proto_version(serverssl, TLS1_2_VERSION));
+        if (current_test_vector->preference == SERVER_PREFERENCE)
+            SSL_set_options(serverssl, SSL_OP_CIPHER_SERVER_PREFERENCE);
+    }
+
+    /* Creating a test connection */
+    TEST_true_or_end(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE));
+
+    /*
+     * Checking that the negotiated group matches our expectation
+     * and must be identical on server- and client-side obviously
+     */
+    negotiated_group_client = SSL_get_negotiated_group(clientssl);
+    negotiated_group_server = SSL_get_negotiated_group(serverssl);
+    if (!TEST_int_eq(negotiated_group_client, negotiated_group_server))
+        goto end;
+    if (!TEST_int_eq(negotiated_group_client, OBJ_sn2nid(current_test_vector->expected_group)))
+        goto end;
+
+    ok = 1;
+
+end:
+    SSL_free(serverssl);
+    SSL_free(clientssl);
+    SSL_CTX_free(server_ctx);
+    SSL_CTX_free(client_ctx);
+    return ok;
+}
+#endif /* !defined(OPENSSL_NO_EC) */
+
+static int tls12groupselection_test(void)
+{
+    int testresult = 1; /* Assume the test will succeed */
+#if !defined(OPENSSL_NO_EC) && !defined(OPENSSL_NO_TLS1_2)
+    unsigned long i;
+
+    /*
+     * Call the code under test, once such that the ssl object is used,
+     * once such that the ctx is usedIf any of the tests fail (= return 0),
+     * the end result will be 0 thanks to multiplication
+     */
+    for (i = 0;
+         i < sizeof(tls12groupselection_tests) / sizeof(tls12groupselection_tests[0]);
```

Comment:
You can use `OSSL_NELEM` for this.

I would also instead use text indexes for this, i.e do this in `setup_tests`:

```` C
    ADD_ALL_TESTS(tls12groupselection_test, OSSL_NELEM(tls12groupselection_tests));
````